tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/reader_moreh_nll_loss_backward_2d.cpp

```diff
@@ -0,0 +1,162 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "debug/dprint.h"  // required in all kernels using DPRINT
+#include "tt_eager/tt_dnn/kernels/dataflow/moreh_common.hpp"
+
+void kernel_main() {
+    uint32_t i = 0;
+    auto target_addr = get_arg_val<uint32_t>(i++);
+    auto weight_addr = get_arg_val<uint32_t>(i++);
+    auto divisor_addr = get_arg_val<uint32_t>(i++);
+    auto output_grad_addr = get_arg_val<uint32_t>(i++);
+    auto ignore_index = static_cast<int32_t>(get_arg_val<uint32_t>(i++));
+    auto num_tiles_per_core = get_arg_val<uint32_t>(i++);
+    auto start_id = get_arg_val<uint32_t>(i++);
+    auto C = get_arg_val<uint32_t>(i++);
+    auto weight_num_tile = get_arg_val<uint32_t>(i++);
+    auto element_size = get_arg_val<uint32_t>(i++);
+
+    constexpr uint32_t cb_output_grad = tt::CB::c_in0;
+    constexpr uint32_t cb_target = tt::CB::c_in1;
+    constexpr uint32_t cb_weight = tt::CB::c_in2;
+    constexpr uint32_t cb_divisor = tt::CB::c_in3;
+
+    constexpr uint32_t cb_tmp_weight = tt::CB::c_intermed0;
+
+    // ublocks size defined in tiles
+    const uint32_t weight_tile_bytes = get_tile_size(cb_weight);
+    const DataFormat weight_data_format = get_dataformat(cb_weight);
+
+    const uint32_t divisor_tile_bytes = get_tile_size(cb_divisor);
+    const DataFormat divisor_data_format = get_dataformat(cb_divisor);
+
+    const uint32_t output_grad_tile_bytes = get_tile_size(cb_output_grad);
+    const DataFormat output_grad_data_format = get_dataformat(cb_output_grad);
+
+    const uint32_t target_tile_bytes = get_tile_size(cb_target);
+
+    constexpr bool target_is_dram = get_compile_time_arg_val(0) == 1;
+    constexpr bool weight_is_dram = get_compile_time_arg_val(1) == 1;
+    constexpr bool divisor_is_dram = get_compile_time_arg_val(2) == 1;
+    constexpr bool output_grad_is_dram = get_compile_time_arg_val(3) == 1;
+
+    const InterleavedAddrGen<target_is_dram> addrg_target = {
+        .bank_base_address = target_addr, .page_size = target_tile_bytes};
+    const InterleavedAddrGenFast<output_grad_is_dram> addrg_output_grad = {
+        .bank_base_address = output_grad_addr,
+        .page_size = output_grad_tile_bytes,
+        .data_format = output_grad_data_format};
+    constexpr uint32_t onetile = 1;
+
+    union {
+        float f;
+        uint32_t u;
+    } one, zero;
+    one.f = 1.0f;
+    zero.f = 0.0f;
```

Comment:
How about using this?
https://github.com/tenstorrent/tt-metal/blob/590d0c2fa142ca0047061366219ec88f740bde11/tt_eager/tt_dnn/kernels/dataflow/moreh_common.hpp#L41C1-L44C3

Reply:
Now use the `Scalar` structure